resources/deliberate-ub.md

```diff
@@ -0,0 +1,42 @@
+When a crate explicitly acknowledges that what it does is UB, but prefers keeping that code over UB-free alternatives (or there are no UB-free alternatives), that is always a concerning sign.
+We should evaluate whether there truly is some use-case here that is not currently achievable in well-defined Rust, and work with crate authors on providing a UB-free solution.
+
+## Known cases of deliberate UB
+
+### Cases related to concurrency
+
+* crossbeam's `AtomicCell` implements an SeqLock, which [is well-known to not be compatible with the C++ memory model](http://www.hpl.hp.com/techreports/2012/HPL-2012-68.pdf).
+  Specifically the problem is [this non-atomic volatile read](https://github.com/crossbeam-rs/crossbeam/blob/5d07fe43540d7f21517a51813acd9332744e90cb/crossbeam-utils/src/atomic/atomic_cell.rs#L980) which can cause data races and hence UB.
+  This would be fine if we either (a) adopted LLVM's handling of memory races (then the problematic read would merely return `undef` instead of UB due to a data race), or (b) added [bytewise atomic memcpy](https://github.com/rust-lang/rfcs/pull/3301) and used that instead of the non-atomic volatile load.
+  This is currently *not* UB in the LLVM IR we generate, due to LLVM's different handling of read-write races and because the code carefully uses `MaybeUninit`.
+* crossbeam's `AtomicCell` also uses the standard library `Atomic*` types to do atomic reads and writes of [*any type* that has the right size](https://github.com/crossbeam-rs/crossbeam/blob/master/crossbeam-utils/src/atomic/atomic_cell.rs#L928-L932).
+  However, doing an `AtomicU32` atomic read on a `(u16, u8)` is unsound because the padding byte can be uninitialized.
+  (Also, pointer provenance is lost.)
+  To fix this we need to be able to perform atomic loads at type `MaybeUninit<u32>`.
```

Comment:
@taiki-e I am curious, given your crate https://github.com/taiki-e/atomic-maybe-uninit/ (which IIUC uses inline assembly to add the missing atomic `MaybeUninit` loads) -- why does crossbeam not use that?

Reply:
> why does crossbeam not use that?

This is because there were several issues. Many of them have already been resolved, and the main remaining issue is that no mechanism (something like `cfg(target_has_atomic)`) is provided by the atomic-maybe-uninit side to handle the lack of portability well (https://github.com/taiki-e/atomic-maybe-uninit/issues/9).

I recently introduced [such a mechanism in another crate](https://github.com/taiki-e/portable-atomic/blob/8d1002fcf62909f3d6f57327b28419597c9d2614/src/utils.rs#L247-L693) and am considering porting it to atomic-maybe-uninit.